File: proj4/src/main/java/org/osgeo/proj4j/parser/Proj4Keyword.java
package org.osgeo.proj4j.parser;

import java.util.*;
import org.osgeo.proj4j.*;

public class Proj4Keyword 
{

  public static final String a = "a";
  public static final String b = "b";
  public static final String f = "f";
  public static final String alpha = "alpha";
  public static final String datum = "datum";
  public static final String ellps = "ellps";
  public static final String es = "es";

  public static final String azi = "azi";
  public static final String k = "k";
  public static final String k_0 = "k_0";
  public static final String lat_ts = "lat_ts";
  public static final String lat_0 = "lat_0";
  public static final String lat_1 = "lat_1";
  public static final String lat_2 = "lat_2";
  public static final String lon_0 = "lon_0";
  public static final String lonc = "lonc";
  public static final String pm = "pm";
  
  public static final String proj = "proj";
  
  public static final String R = "R";
  public static final String R_A = "R_A";
  public static final String R_a = "R_a";
  public static final String R_V = "R_V";
  public static final String R_g = "R_g";
  public static final String R_h = "R_h";
  public static final String R_lat_a = "R_lat_a";
  public static final String R_lat_g = "R_lat_g";
  public static final String rf = "rf";
  
  public static final String south = "south";
  public static final String to_meter = "to_meter";
  public static final String towgs84 = "towgs84";
  public static final String units = "units";
  public static final String x_0 = "x_0";
  public static final String y_0 = "y_0";
  public static final String zone = "zone";
  
  public static final String title = "title";
  public static final String nadgrids = "nadgrids";
  public static final String no_defs = "no_defs";
  public static final String wktext = "wktext";


  private static Set<String> supportedParams = null;
  
  public static synchronized Set supportedParameters()
  {
    if (supportedParams == null) {
      supportedParams = new TreeSet<String>();
      
      supportedParams.add(a);
      supportedParams.add(rf);
      supportedParams.add(f);
      supportedParams.add(alpha);
      supportedParams.add(es);
      supportedParams.add(b);
      supportedParams.add(datum);
      supportedParams.add(ellps);
      
      supportedParams.add(R_A);
   
      supportedParams.add(k);
      supportedParams.add(k_0);
      supportedParams.add(lat_ts);
      supportedParams.add(lat_0);
      supportedParams.add(lat_1);
      supportedParams.add(lat_2);
      supportedParams.add(lon_0);
      supportedParams.add(lonc);
      
      supportedParams.add(x_0);
      supportedParams.add(y_0);

      supportedParams.add(proj);
      supportedParams.add(south);
      supportedParams.add(towgs84);
      supportedParams.add(to_meter);
      supportedParams.add(units);
      supportedParams.add(zone);
      
      supportedParams.add(title);       // no-op
      supportedParams.add(no_defs);     // no-op
      supportedParams.add(wktext);      // no-op
      supportedParams.add(nadgrids);    // no-op for now

    }
    return supportedParams;
  }
  
  public static boolean isSupported(String paramKey)
  {
    return supportedParameters().contains(paramKey);
  }
  
  public static void checkUnsupported(String paramKey)
  {
    if (! isSupported(paramKey)) {
      throw new UnsupportedParameterException(paramKey + " parameter is not supported");
    }
  }
  
  public static void checkUnsupported(Collection params)
  {
    for (Object s : params) {
      checkUnsupported((String) s);
    }
  }
}

File: proj4/src/main/scala/geotrellis/proj4/parser/Proj4Keyword.scala
/*
 * Copyright (c) 2014 Azavea.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.proj4.parser

object Proj4Keyword {

  val a = "a"
  val b = "b"
  val f = "f"
  val alpha = "alpha"
  val datum = "datum"
  val ellps = "ellps"
  val es = "es"

  val azi = "azi"
  val k = "k"
  val k_0 = "k_0"
  val lat_ts = "lat_ts"
  val lat_0 = "lat_0"
  val lat_1 = "lat_1"
  val lat_2 = "lat_2"
  val lon_0 = "lon_0"
  val lonc = "lonc"
  val pm = "pm"

  val proj = "proj"

  val R = "R"
  val R_A = "R_A"
  val R_a = "R_a"
  val R_V = "R_V"
  val R_g = "R_g"
  val R_h = "R_h"
  val R_lat_a = "R_lat_a"
  val R_lat_g = "R_lat_g"
  val rf = "rf"

  val south = "south"
  val to_meter = "to_meter"
  val towgs84 = "towgs84"
  val units = "units"
  val x_0 = "x_0"
  val y_0 = "y_0"
  val zone = "zone"

  val title = "title"
  val nadgrids = "nadgrids"
  val no_defs = "no_defs"
  val wktext = "wktext"

  private val supportedParams = 
    Set(
      a, rf, f, alpha, es, b, datum, ellps,
      R_A,
      k, k_0, lat_ts, lat_0, lat_1, lat_2, lon_0, lonc,
      x_0, y_0,
      proj, south, towgs84, to_meter, units, zone,
      title, no_defs, wktext, nadgrids
    )

  def isSupported(paramKey: String): Boolean =
    supportedParams.contains(paramKey)

  /** Returns list of unsupported keys if is not valid, otherwise None */
  def invalidKeys(keySet: Set[String]): Option[List[String]] = {
    val l = keySet.filter(!isSupported(_)).toList
    if(l.size != 0)
      Some(l)
    else
      None
  }
}
